backends/ubpf/ubpfDeparser.cpp

```diff
@@ -0,0 +1,235 @@
+                            widthToEmit > 64 ? bytes - i - 1 : i, // reversed order for wider fields
+                            bitsToWrite,
+                            8 + alignment - bitsToWrite,
+                            program->byteVar.c_str());
+                }
+                builder->endOfStatement(true);
+                left -= bitsToWrite;
+            }
+            alignment = (alignment + bitsToWrite) % 8;
+
+        }
+
+        builder->emitIndent();
+        builder->appendFormat("%s += %d", program->offsetVar.c_str(),
+                              widthToEmit);
+        builder->endOfStatement(true);
+        builder->newline();
+    }
+
+    void UBPFDeparserTranslationVisitor::compileEmit(const IR::Vector<IR::Argument> *args) {
+        BUG_CHECK(args->size() == 1, "%1%: expected 1 argument for emit", args);
+
+        auto expr = args->at(0)->expression;
+        auto type = typeMap->getType(expr);
+        auto ht = type->to<IR::Type_Header>();
+        if (ht == nullptr) {
+            ::error("Cannot emit a non-header type %1%", expr);
+            return;
+        }
+
+        auto program = deparser->program;
+        builder->emitIndent();
+        builder->append("if (");
+        visit(expr);
+        builder->append(".ebpf_valid) ");
+        builder->blockStart();
+
+        unsigned alignment = 0;
+        for (auto f : ht->fields) {
+            auto ftype = typeMap->getType(f);
+            auto etype = UBPFTypeFactory::instance->create(ftype);
+            auto et = dynamic_cast<EBPF::IHasWidth *>(etype);
+            if (et == nullptr) {
+                ::error("Only headers with fixed widths supported %1%", f);
+                return;
+            }
+            compileEmitField(expr, f->name, alignment, etype);
+            alignment += et->widthInBits();
+            alignment %= 8;
+        }
+
+        builder->blockEnd(true);
+
+    }
+
+    bool UBPFDeparserTranslationVisitor::preorder(const IR::MethodCallExpression *expression) {
+
+        auto mi = P4::MethodInstance::resolve(expression,
+                                              deparser->program->refMap,
+                                              deparser->program->typeMap);
+        auto extMethod = mi->to<P4::ExternMethod>();
+
+        if (extMethod != nullptr) {
+            auto decl = extMethod->object;
+            if (decl == deparser->packet_out) {
+                if (extMethod->method->name.name == p4lib.packetOut.emit.name) {
+                    compileEmit(extMethod->expr->arguments);
+                    return false;
+                }
+            }
+        }
+
+        ::error("Unexpected method call in deparser %1%", expression);
+        return false;
+    }
+
+    bool UBPFDeparser::build() {
+        auto pl = controlBlock->container->type->applyParams;
+        if (pl->size() != 2) {
+            ::error("Expected deparser to have exactly 2 parameters");
```

Comment:
You should supply some IR::Node argument to error.

Reply:
Done.